salt/config/__init__.py

```diff
@@ -3360,13 +3422,18 @@ def get_id(opts, cache_minion_id=False):
         log.debug('Guessing ID. The id can be explicitly set in {0}'
                   .format(os.path.join(salt.syspaths.CONFIG_DIR, 'minion')))
 
-    newid = salt.utils.network.generate_minion_id()
+    if opts.get('id_function'):
+        newid = eval_id_function(opts)
+    else:
+        newid = salt.utils.network.generate_minion_id()
```

Comment:
Hmm. I would not do this, as here is a potential bug: in your `eval_id_function` you can possibly [return an empty value](https://github.com/saltstack/salt/pull/41619/files#diff-0ddbe12a41fee1a8e552b35086f918ddR3339) and then we have a trouble. Instead, I would do something like:
```python

def find_minion_id(opts):
    '''
    This is your renamed "eval_id_function" that calls 
    salt.utils.network.generate_minion_id() if no value.
    '''
    ...
    # All your "eval_id_function" code, but do not return anywhere.
    ...
    return opts.get('id') or salt.utils.network.generate_minion_id()

...
def get_id(opts, cache_minion_id=False):
    new_id = find_minion_id(opts)  # let the function alone figure out
                                   # required params from opts
    ...
```